push/ups_integration_with_openshift.md

```diff
@@ -0,0 +1,190 @@
+
+A future change to this could look at adding support for individual 'developer' users.
+This would be required if running UPS as a shared service.
+That would most likely require changes to UPS to call OpenShift to check user permissions based on the token header, and only showing that developers Push Applications.
+
+These changes will mean UPS will no longer have a dependency on Keycloak when running in OpenShift.
+Also, the developer doesn't have to know about or be concerned with where or how the auth provider is setup and configured for UPS.
+
+## OpenShift UI Integration
+
+To address the third story, a number of UI screens and integration points will be added to the OpenShift UI & the UPS APB.
+
+Push Variant credentials will be stored in OpenShift as a Secret.
+The Secret will have:
+
+* a label to identify it as storing Variant credentials
+* an annotation to link to a specific Mobile Client `aerogear.org/mobile-client-id`
+* an annotation to store the ID of the Variant in UPS `aerogear.org/variant-id`
+* an annotation to store the Variant config from UPS `aerogear.org/variant-config-json`
+* a generated name in the format `variant-<variant-type>-<timestamp>`
+
+Example Variant Secret:
+
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  annotations:
+    aerogear.org/mobile-client-id: 334e1333b5d4-e114-4332-8332-337c4111
+    aerogear.org/variant-id: 337c42c5-f444-4762-89ef-334e1684b5d4
+    aerogear.org/variant-config-json: >-
+      {"variantId":"some-variant-id","variantSecret":"some-secret","senderId":some-sender-id"}
+  labels:
+    variant: true
+  name: variant-android-334e1333b5d4-e114-4332-8332-337c4111-151222332230
+  namespace: myproject
+type: Opaque
+data:
+  googleKey: >-
+    \c3VwZXJzZWNyZXQK
+```
+
+A sidecar will be used to sync these Push Variant credentials from OpenShift to UPS.
+The Variant ID and Config will be set by the sidecar after creation of the Variant in UPS.
+OpenShift will be the source of truth for Variants linked to Mobile Clients.
+The sycning process in the sidecar will need to consider:
+
+* Syncing *From* OpenShift *to* UPS
+  * New Secrets mean a New Variant is created
+  * Updates to Secrets overwrite the Variant
+  * Deleting a Secret will delete the Variant
+  * Deleting a Variant directly in UPS will cause it to be recreated
+* Syncing *From* UPS *to* OpenShift
+  * Variant ID will be annotated on the Secret after creation
+  * Variant Config will be annotated on the Secret and kept up to date (UPS is the source of truth for this config)
+
+Here is an example workflow showing the UI touchpoints:
+
+* developer creates a Cordova App (MobileClient) in OpenShift
+* developer provisions UPS from the Catalog
+* developer sets up the FCM/Android credentials for push for the Cordova App via *new* UI screens in OpenShift
+* this creates a Secret to store those credentials
```

Comment:
we actually don't need to store the credentials in the secret if we are storing the returned variant definition.

Reply:
True, however, i think it's best to store them in case we need to re-create/update the Variant.
If the sidecar is watching this resource for changes, it can pick up on a new uploaded cert or changed key, then propagate that change to UPS.
This will mean we can keep the logic for managing variants in a single place (i.e. in the sidecar).
A variant will only every get created by the sidecar.
If we say a 'bind' to the ups apb sets up the variant, then the 'bind' is just creating a Secret with the correct label and data for the sidecar to work with.